pkg/lib/oauth/handler/handler_token.go

```diff
@@ -290,6 +290,24 @@ func (h *TokenHandler) app2appGetDeviceKeyJWKVerified(jwt string) (jwk.Key, erro
 	return key, nil
 }
 
+func (h *TokenHandler) rotateDeviceSecretIfNeeded(offlineGrant *oauth.OfflineGrant, resp protocol.TokenResponse) (*oauth.OfflineGrant, error) {
+	if offlineGrant.DeviceSecretHash == "" {
+		// No device secret, no rotation needed.
+		return offlineGrant, nil
+	}
```

Comment:
According to the spec, we should issue device secret if it is invalid, or absent, given that the authorization code has the scope.

It seems that we haven't handle `grant_type=refresh_token` yet.

I think we should rotate the device secret every time the refresh token is used, this includes:
- grant_type=refresh_token, that is, when the refresh token is used to refresh an access token
- reauthenticate
- refresh ID token with our custom grant type